File: 42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0, r = height.size() - 1, ans = 0, lm = 0, rm = 0;
        while(l < r) {
            if(height[l] < height[r]) {
                if(height[l] > lm)
                    lm = height[l];
                else
                    ans += lm - height[l];
                l++;
            }
            else {
                if(height[r] > rm)
                    rm = height[r];
                else
                    ans += rm - height[r];
                r--;
            }
        }
        return ans;
    }
};

File: 42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r, lmax, rmax = 0, len(height) - 1, 0, 0
        ans = 0

        while(l < r):
            if(height[l] < height[r]):
                if(height[l] > lmax):
                    lmax = height[l]
                else:
                    ans += lmax - height[l]
                l += 1
            
            else:
                if(height[r] > rmax):
                    rmax = height[r]
                else:
                    ans += rmax - height[r]
                r -= 1
        

        return ans
        
